flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/plan/logical/operators.scala

```diff
@@ -0,0 +1,309 @@
+    }
+
+    def validateGroupingExpression(expr: Expression): Unit = {
+      if (!expr.dataType.isKeyType) {
+        failValidation(
+          s"expression $expr cannot be used as a grouping expression " +
+            "because it's not a valid key type")
+      }
+    }
+    resolvedAggregate
+  }
+}
+
+case class Union(left: LogicalNode, right: LogicalNode) extends BinaryNode {
+  override def output: Seq[Attribute] = left.output
+
+  override def toRelNode(relBuilder: RelBuilder): RelBuilder = {
+    left.toRelNode(relBuilder)
+    right.toRelNode(relBuilder)
+    relBuilder.union(true)
+  }
+
+  override def validate(tableEnv: TableEnvironment): LogicalNode = {
+    val resolvedUnion = super.validate(tableEnv).asInstanceOf[Union]
+    if (left.output.length != right.output.length) {
+      failValidation(s"Union two table of different column sizes:" +
+        s" ${left.output.size} and ${right.output.size}")
+    }
+    val sameSchema = left.output.zip(right.output).forall { case (l, r) =>
+      l.dataType == r.dataType && l.name == r.name }
+    if (!sameSchema) {
+      failValidation(s"Union two table of different schema:" +
+        s" [${left.output.map(a => (a.name, a.dataType)).mkString(", ")}] and" +
+        s" [${right.output.map(a => (a.name, a.dataType)).mkString(", ")}]")
+    }
+    resolvedUnion
+  }
+}
+
+case class Join(
+    left: LogicalNode,
+    right: LogicalNode,
+    joinType: JoinType,
+    condition: Option[Expression]) extends BinaryNode {
+
+  override def output: Seq[Attribute] = {
+    joinType match {
+      case JoinType.INNER => left.output ++ right.output
+      case j => throw new ValidationException(s"Unsupported JoinType: $j")
+    }
+  }
+
+  override def toRelNode(relBuilder: RelBuilder): RelBuilder = {
+    joinType match {
+      case JoinType.INNER =>
+        left.toRelNode(relBuilder)
+        right.toRelNode(relBuilder)
+        relBuilder.join(JoinRelType.INNER,
+          condition.map(_.toRexNode(relBuilder)).getOrElse(relBuilder.literal(true)))
+      case _ =>
+        throw new ValidationException(s"Unsupported JoinType: $joinType")
+    }
+  }
+
+  private def ambiguousName: Set[String] =
+    left.output.map(_.name).toSet.intersect(right.output.map(_.name).toSet)
+
+  override def validate(tableEnv: TableEnvironment): LogicalNode = {
+    if (tableEnv.isInstanceOf[StreamTableEnvironment]) {
+      failValidation(s"Join on stream tables is currently not supported.")
+    }
+
+    val resolvedJoin = super.validate(tableEnv).asInstanceOf[Join]
+    if (!resolvedJoin.condition.forall(_.dataType == BOOLEAN_TYPE_INFO)) {
+      failValidation(s"filter expression ${resolvedJoin.condition} is not a boolean")
+    } else if (!ambiguousName.isEmpty) {
+      failValidation(s"join relations with ambiguous names: ${ambiguousName.mkString(", ")}")
+    }
+    resolvedJoin
+  }
+}
+
+case class CatalogNode(
+    tableName: String,
+    table: CTable,
```

Comment:
As mentioned in `TableEnvironment`, I think we do not need the Calcite `Table` and `RelDataTypeFactory` if we request the `rowType` directly from the `TableEnvironment`.